modules/problem_util.js

```diff
@@ -1,24 +1,59 @@
+'use strict';
+
 const path          = require('path'); 
 const fs            = require('fs');
 const yaml          = require('js-yaml');
+const datastore     = require('./datastore.js');
 
 const PROBLEMS_DIR  = path.join(__dirname, '../problems');
+const PROBLEMS_KIND = 'Problem';
 
 /**
- * Placeholder function that will be replaced once Datastore functionality
- * is ready.
- *
- * When implementing this functionality once Datastore is ready, make sure to
- * guarantee idempotency as it is likely that mass updating problems will be
- * a common operation.
- * Ex   : User added 20 new problems but there is still 5 old problems in /problems
- *        that should not create duplicate problems.
- *
- * Separated into its own function in case if we want to support adding problems
- * by other means than files in /problems such as using form in the web application.
+ * Adds a new problem to the Datastore given a problem object.
+ * Asserts that the problem is valid using problemIsValid
+ * and guarantees idempotency.
+ **/
+async function addProblem(problemObject) {
+  if (!problemIsValid(problemObject))
+    return;
+
+  const key = await generateKey(problemObject);
+  datastore.store(key, problemObject);
+}
+
+/**
+ * Generates key given a problem object to guarantee idempotency.
+ * If a problem with the same id already exists, it will give the
+ * entity's key such that it could be updated.
+ * Otherwise, PROBLEMS_KIND is simply returned, indicating
+ * that a new entity should be created.
+ **/
+async function generateKey(problemObject) {
+  const query = datastore
+    .createQuery(PROBLEMS_KIND)
+    .filter('id', '=', problemObject.id);
+
+  const [problems] = await datastore.runQuery(query);
```

Comment:
for unique IDs this is okay. in general though, we should account for pagination.
`const [problems,info] = await datastore.runQuery(query);`
then paginate with info.

this is where it would be handy to have a helper function that can do pagination for general queries.

Reply:
Thanks for the feedback!
I decided to do that for now as we only want unique id in this case.

The pods have also talked about features such as pagination and we decided to do it post-MVP for now as we still have higher-priority features.